zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,256 @@
+public class ZookeeperEndpointGroup implements EndpointGroup {
+
+    private static final Logger logger = LoggerFactory.getLogger(ZookeeperEndpointGroup.class);
+
+    private static final int MAX_RETRY_DELAY = 60 * 1000; //one minute
+    private int retryDelay = 1000; //start with one second
+    private final String zkConnectionStr;
+    private final String zNode;
+    private final int sessionTimeout;
+    private final ZkNodeValueConverter converter;
+    private List<Endpoint> endpointList;
+    private ZooKeeper zooKeeper;
+    private byte[] prevData;
+    private CompletableFuture<ZooKeeper> zkHandleProxy = new CompletableFuture<>();
+
+    /**
+     * Creates a {@link ZookeeperEndpointGroup}.
+     * @param zkConnectionStr A connection string containing a comma
+     *                          separated list of host:port pairs , each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     * @param converter a function to convert zNode value to a List of {@link Endpoint}
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout,
+                                  ZkNodeValueConverter converter) {
+        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
+        this.zNode = requireNonNull(zNode, "zNode");
+        this.converter = requireNonNull(converter, "converter");
+        this.sessionTimeout = sessionTimeout;
+        doConnect();
+    }
+
+    /**
+     * Create a Zookeeper endpoint group with a {@link DefaultZkNodeValueConverter}.
+     * @param zkConnectionStr A connection string containing a comma
+     *                          separated list of host:port pairs , each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultZkNodeValueConverter());
+    }
+
+    /**
+     *<h3>note:</h3>.
+     *This method could be invoked in the callback thread instead of the main thread,
+     *The resetting operation of the Zookeeper client handler could cause a race condition between the
+     *two threads,we need to use a proxy to protect it
+     */
+    private void doConnect() {
+        try {
+            zooKeeper = new ZooKeeper(zkConnectionStr, sessionTimeout, new ZkWatcher());
+
+            Stat stat = zkHandleProxy.get().exists(zNode, true);
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zkHandleProxy.get().getData(zNode, false, null);
+                endpointList = converter.convert(nodeData);
+            }
+        } catch (Exception e) {
+            throw new EndpointGroupException(
```

Comment:
no, this is just a `connect` operation,  when you invoke this method at the application start time , it will just throw exceptions(and let the user know something is going wrong and their application can not start normally). but when you call it in the Watcher's call back( for reconnecting the server after a network partition  ) ,it will catch the exception and do the retry routine.